lib/native/source/opensmt/api.patch

```diff
@@ -0,0 +1,537 @@
++#include <mutex>
+ 
+ typedef int32_t  word;
+ typedef uint32_t uword;
+@@ -71,6 +72,7 @@ class FastRational
+ {
+     class mpqPool
+     {
++        std::mutex cs; // We use a mutex to make the pool thread-safe
+         std::stack<mpq_class> store; // uses deque as storage to avoid realloc
+         std::stack<mpq_ptr, std::vector<mpq_ptr>> pool;
+     public:
+diff --git a/src/logics/ArithLogic.cc b/src/logics/ArithLogic.cc
+index 8609ef55..11813348 100644
+--- a/src/logics/ArithLogic.cc
++++ b/src/logics/ArithLogic.cc
+@@ -807,7 +807,7 @@ PTRef ArithLogic::mkMod(vec<PTRef> && args) {
+     PTRef dividend = args[0];
+     PTRef divisor = args[1];
+ 
+-    if (not isNumConst(divisor)) { throw OsmtApiException("Divisor must be constant in linear logic"); }
++    if (not isNumConst(divisor)) { throw LANonLinearException("Divisor must be constant in linear logic"); }
+     if (isZero(divisor)) { throw ArithDivisionByZeroException(); }
+     if (isOne(divisor) or isMinusOne(divisor)) { return getTerm_IntZero(); }
+     if (isConstant(dividend)) {
+@@ -829,7 +829,7 @@ PTRef ArithLogic::mkIntDiv(vec<PTRef> && args) {
+     assert(args.size() == 2);
+     PTRef dividend = args[0];
+     PTRef divisor = args[1];
+-    if (not isConstant(divisor) and not isConstant(dividend)) { throw LANonLinearException("Divisor or dividend must be constant in linear logic"); }
++    if (not isConstant(divisor)) { throw LANonLinearException("Divisor or dividend must be constant in linear logic"); }
+     if (isZero(divisor)) { throw ArithDivisionByZeroException(); }
+     if (isOne(divisor)) { return dividend; }
+     if (isMinusOne(divisor)) { return mkNeg(dividend); }
+@@ -1177,7 +1177,7 @@ std::string ArithLogic::printTerm_(PTRef tr, bool ext, bool safe) const
+         char* tmp_str;
+         opensmt::stringToRational(tmp_str, sym_store.getName(getPterm(tr).symb()));
+         opensmt::Number v(tmp_str);
+-        if (!isNonNegNumConst(tr)) {
++        if (!isNonNegative(v)) {
+             v.negate();
+             is_neg = true;
+         }
+diff --git a/src/logics/ArithLogic.h b/src/logics/ArithLogic.h
+index 6f2669f9..fa8a7e26 100644
+--- a/src/logics/ArithLogic.h
++++ b/src/logics/ArithLogic.h
+@@ -4,7 +4,7 @@
+ #include "Number.h"
+ #include <numeric>
+ 
+-class LANonLinearException : std::runtime_error
++class LANonLinearException : public std::runtime_error
+ {
+     std::string msg;
+ public:
+@@ -417,4 +417,4 @@ public:
+     bool operator ()  (PTRef x_, PTRef y_) const;
+ 
+ };
+-#endif
+\ No newline at end of file
++#endif
+diff --git a/src/logics/FunctionTools.h b/src/logics/FunctionTools.h
+index fc9dbfb7..943a466d 100644
+--- a/src/logics/FunctionTools.h
++++ b/src/logics/FunctionTools.h
+@@ -33,7 +33,7 @@ class TemplateFunction {
+     FunctionSignature signature;
+     PTRef tr_body;
+ 
+-    inline static std::string_view constexpr template_arg_prefix = ".arg";
++    inline static constexpr std::string_view template_arg_prefix = ".arg";
```

Comment:
Is there a reason for changing the order of modifiers in OpenSMT API? This makes the patch quite complex. We should prefer to only patch a minimum number of lines and - if possible - let the OpenSMT team fix other issues in their code.

Reply:
We can definitely fix these problems in OpenSMT itself. Could you open issues for these findings?

Reply:
Hello Martin,
I'll split the patch up and then open a pull request for OpenSMT so you can pick and choose the changes you want to add to OpenSMT.

Reply:
I've now opened a pull request for OpenSMT: https://github.com/usi-verification-and-security/opensmt/pull/646